File: CrossDomainPGELLAv1.0/updatePGELLA.m
function [modelPGELLA]=updatePGELLA(modelPGELLA,taskId,ObservedTasks,HessianArray,ParameterArray,Tasks)
% This function updates the PG-ELLA model, i.e., L and S

% Updating L
[AllTasks]=find(ObservedTasks);
% From the observed tasks, determine which groups we have .. 
for l=1:length(AllTasks)
    GroupsPresentD(l,:)=Tasks(AllTasks(l)).param.Group; 
end
GroupsPresent=unique(GroupsPresentD);
for z=1:length(GroupsPresent) % sum over the goals ... 
    % Determining from observed tasks which ones are allowed 
    [allowedTaskIndexD]=getAllowedTasks(GroupsPresent(z,:),ObservedTasks,Tasks);
    allowedTaskIndexG{z}=allowedTaskIndexD(find(allowedTaskIndexD)); 
end

% Having determined the allowed tasks, compute the derivative 
sumAllOne=0;
TasksforGroupz=allowedTaskIndexG{find(GroupsPresent==Tasks(taskId).param.Group)};
sum=zeros(size(modelPGELLA(Tasks(taskId).param.Group).L));
Tg=length(TasksforGroupz);

for i=1:Tg
    sum=sum-2*HessianArray(TasksforGroupz(i)).D*ParameterArray(TasksforGroupz(i)).alpha*modelPGELLA(1).S(:,i)' ...
            +2*HessianArray(TasksforGroupz(i)).D*modelPGELLA(Tasks(taskId).param.Group).L*modelPGELLA(1).S(:,i)*modelPGELLA(1).S(:,i)' ...
            +2*modelPGELLA(1).mu_two*modelPGELLA(Tasks(taskId).param.Group).L; 
end

modelPGELLA(Tasks(taskId).param.Group).L = (modelPGELLA(Tasks(taskId).param.Group).L+modelPGELLA(1).learningRateL*1./Tg*sum);

% Update s_{taskId} using LASSO

% Determine which group taskId belongs to 
Dsqrt = real(HessianArray(taskId).D^.5);
target = Dsqrt*ParameterArray(taskId).alpha;
dictTransformed = Dsqrt*modelPGELLA(Tasks(taskId).param.Group).L;

s = full(mexLasso(target,dictTransformed,struct('lambda',modelPGELLA(1).mu_one/2)));
modelPGELLA(1).S(:,taskId)=s; 

File: turtle_experiments/src/updatePGELLA.py
# -*- coding: utf-8 -*-
"""
Created on Tue Aug 18 13:50:37 2015

@author: brownyoda
"""

import numpy as np
import spams


def updatePGELLA(ELLAmodel, taskId, ObservedTasks, HessianArray, ParameterArray):

    #------------------------------------------------------------------------
    # Update L -- Tasks[taskId].param.Group --> Know which Group ..
    #------------------------------------------------------------------------
    summ = np.zeros(np.shape(ELLAmodel.L))
    allowedTask = np.nonzero(ObservedTasks == 1)[0]
    Tg = np.sum(ObservedTasks)

    for i in range(int(Tg)):
        S = ELLAmodel.S[:,allowedTask[i]].reshape(1,1)
        S_p = S.conj().T
        print "shape 1: ", np.shape((2 * HessianArray[allowedTask[i]].D))
        print "shape 2: ", np.shape(np.dot(ParameterArray[allowedTask[i]].alpha, S_p))
        print "2nd term:", np.dot((2 * HessianArray[allowedTask[i]].D), np.dot(ParameterArray[allowedTask[i]].alpha, S_p)).reshape(np.shape(ELLAmodel.L))
        print "3rd term:", np.dot((2 * HessianArray[allowedTask[i]].D), np.dot(ELLAmodel.L, np.dot(ELLAmodel.S[:,allowedTask[i]], ELLAmodel.S[:,allowedTask[i]].conj().T)))
        print "4th term:", 2 * ELLAmodel.mu_two * ELLAmodel.L
        summ = summ - np.dot((2 * HessianArray[allowedTask[i]].D), np.dot(ParameterArray[allowedTask[i]].alpha, ELLAmodel.S[:,allowedTask[i]].conj().T)).reshape(np.shape(ELLAmodel.L)) \
                    + np.dot((2 * HessianArray[allowedTask[i]].D), np.dot(ELLAmodel.L, np.dot(ELLAmodel.S[:,allowedTask[i]], ELLAmodel.S[:,allowedTask[i]].conj().T))) \
                    + 2 * ELLAmodel.mu_two * ELLAmodel.L

    ELLAmodel.L = (ELLAmodel.L - ELLAmodel.learningRate * 1.0 / Tg * summ)
    #------------------------------------------------------------------------
    # Update s_{taskId} using LASSO
    #------------------------------------------------------------------------
    # Determine which group taskId belongs to
    Dsqrt = HessianArray[taskId].D ** .5
    target = np.dot(Dsqrt, ParameterArray[taskId].alpha)
    dictTransformed = np.dot(Dsqrt, ELLAmodel.L)

    '''
    lasso(X,D= None,Q = None,q = None,return_reg_path = False,L= -1,lambda1= None,lambda2= 0.,
                 mode= spams_wrap.PENALTY,pos= False,ols= False,numThreads= -1,
                 max_length_path= -1,verbose=False,cholesky= False):
    '''
    s = spams.lasso(np.asfortranarray(target, dtype=np.float64), D = np.asfortranarray(dictTransformed, dtype=np.float64),
                            Q = None, q = None, return_reg_path = False, L = -1, lambda1 = ELLAmodel.mu_one / 2.0, lambda2 = 0., verbose = False, mode = 2)

    ELLAmodel.S[:, taskId] = np.asarray(s.todense())

    print "ELLA Model: ", ELLAmodel.S

    return ELLAmodel